File: app/src/main/java/droiddevelopers254/droidconke/views/fragments/DayOneFragment.java
package droiddevelopers254.droidconke.views.fragments;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import droiddevelopers254.droidconke.R;
import droiddevelopers254.droidconke.adapters.SessionTimeAdapter;
import droiddevelopers254.droidconke.adapters.SessionsAdapter;
import droiddevelopers254.droidconke.models.SessionTimeModel;
import droiddevelopers254.droidconke.models.SessionsModel;
import droiddevelopers254.droidconke.utils.ItemClickListener;
import droiddevelopers254.droidconke.viewmodels.DayOneViewModel;
import droiddevelopers254.droidconke.views.activities.SessionViewActivity;

public class DayOneFragment extends Fragment {
    SessionsAdapter sessionsAdapter;
    List<SessionsModel> sessionsModelList = new ArrayList<>();
    List<SessionTimeModel> sessionTimeModelList = new ArrayList<>();
    List<String> sessionIds = new ArrayList<>();
    static RecyclerView.LayoutManager mLayoutManager;
    DayOneViewModel dayOneViewModel;
    @BindView(R.id.sessionsRv)
    RecyclerView sessionsRv;
    Unbinder unbinder;
    SessionTimeAdapter sessionTimeAdapter;
    boolean isStarred;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_day_one, container, false);

        dayOneViewModel = ViewModelProviders.of(this).get(DayOneViewModel.class);
        unbinder = ButterKnife.bind(this, view);

        initView();

        dayOneViewModel.getDayOneSessions();
        //observe live data emitted by view model
        dayOneViewModel.getSessions().observe(this,sessionsState -> {
            if(sessionsState.getSessionsModel() != null){
                sessionsModelList= sessionsState.getSessionsModel();

                //set the data on the adapter
                sessionsAdapter.setSessionsAdapter(sessionsModelList);

            }else {
                handleError(sessionsState.getDatabaseError());
            }
        });

        return view;
    }

    private void handleError(String databaseError) {
        Toast.makeText(getActivity(),databaseError,Toast.LENGTH_SHORT).show();
    }

    private void initView() {
        sessionsAdapter = new SessionsAdapter(getActivity(), sessionsModelList, "day_one");
        mLayoutManager = new LinearLayoutManager(getActivity());
        sessionsRv.setLayoutManager(mLayoutManager);
        sessionsRv.setItemAnimator(new DefaultItemAnimator());
        sessionsRv.setAdapter(sessionsAdapter);
        sessionsRv.addOnItemTouchListener(new ItemClickListener(getContext(), sessionsRv, new ItemClickListener.ClickListener() {
            @Override
            public void onClick(View view, int position) {
                Intent intent = new Intent(getContext(), SessionViewActivity.class);
                intent.putExtra("sessionId", sessionsModelList.get(position).getId());
                intent.putExtra("dayNumber", "day_one");
                intent.putExtra("starred", sessionsModelList.get(position).getIsStarred());
                intent.putIntegerArrayListExtra("speakerId", sessionsModelList.get(position).getSpeaker_id());
                intent.putExtra("roomId", sessionsModelList.get(position).getRoom_id());
                intent.putExtra("sessionName",sessionsModelList.get(position).getTitle());
                intent.putExtra("sessionUrl",sessionsModelList.get(position).getUrl());
                intent.putExtra("sessionColor",sessionsModelList.get(position).getSession_color());
                intent.putExtra("photoUrl",sessionsModelList.get(position).getPhotoUrl());
                startActivity(intent);
            }

            @Override
            public void onLongClick(View view, int position) {

            }
        }));

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }


}

File: app/src/main/java/droiddevelopers254/droidconke/views/fragments/DayOneFragment.kt
package droiddevelopers254.droidconke.views.fragments

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import droiddevelopers254.droidconke.R
import droiddevelopers254.droidconke.adapters.SessionsAdapter
import droiddevelopers254.droidconke.models.SessionsModel
import droiddevelopers254.droidconke.utils.ItemClickListener
import droiddevelopers254.droidconke.utils.nonNull
import droiddevelopers254.droidconke.utils.observe
import droiddevelopers254.droidconke.viewmodels.DayOneViewModel
import droiddevelopers254.droidconke.views.activities.SessionViewActivity
import kotlinx.android.synthetic.main.fragment_day_one.*
import kotlinx.android.synthetic.main.fragment_day_one.view.*
import org.jetbrains.anko.toast
import java.util.*

class DayOneFragment : Fragment() {

    internal var sessionsModelList: List<SessionsModel> = ArrayList()
    lateinit var dayOneViewModel: DayOneViewModel
    lateinit var sessionsAdapter: SessionsAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_day_one, container, false)

        dayOneViewModel = ViewModelProviders.of(this).get(DayOneViewModel::class.java)
        sessionsAdapter = SessionsAdapter(activity!!, sessionsModelList, "day_one")
        val sessionsRv = view.sessionsRv

        initView(sessionsRv)

        dayOneViewModel.getDayOneSessions()

        //observe live data emitted by view model
        observeLiveData()
        return view
    }

    private fun observeLiveData() {
        dayOneViewModel.getSessionsResponse().nonNull().observe(this){
            sessionsModelList = it
            sessionsAdapter.setSessionsAdapter(sessionsModelList)
            loginProgressBar.visibility = View.GONE

        }
        dayOneViewModel.getSessionsError().nonNull().observe(this){
            handleError(it)
        }
    }

    private fun handleError(databaseError: String?) {
        activity?.toast(databaseError.toString())
    }

    private fun initView(sessionsRv: RecyclerView) {
        val layoutManager = LinearLayoutManager(activity)
        sessionsRv.layoutManager = layoutManager
        sessionsRv.itemAnimator = DefaultItemAnimator()
        sessionsRv.adapter = sessionsAdapter
        sessionsRv.addOnItemTouchListener(ItemClickListener(context!!, sessionsRv, object : ItemClickListener.ClickListener {

            override fun onLongClick(view: View?, position: Int) {
            }

            override fun onClick(view: View, position: Int) {
                val intent = Intent(context, SessionViewActivity::class.java)
                intent.putExtra("sessionId", sessionsModelList[position].id)
                intent.putExtra("dayNumber", "day_one")
                intent.putExtra("starred", sessionsModelList[position].starred)
                intent.putIntegerArrayListExtra("speakerId", sessionsModelList[position].speaker_id)
                intent.putExtra("roomId", sessionsModelList[position].room_id)
                startActivity(intent)
            }


        }))

    }

}
